### PR TITLE
Add backend support for notifications

### LINKED ISSUE #238: Add notifications support

### PR DESCRIPTION
**Pull request by BogdanCDev**
## Summary

This pull request introduces notification support for the Voyageur Android app, including permission handling and a helper class for managing notifications. This feature enhances user experience by notifying users about critical events, improving overall app usability and engagement.

## Changes

- **Screens/UI:** Updated MainActivity to include logic for triggering notifications 
- **Logic/Controllers:** 
  - Added utility methods:
    - createNotificationChannel: Sets up notification channels for Android O and above.
    - showNotification: Builds and displays notifications with configurable priority.
    - showNoInternetNotification: Specialised method for no internet notifications.
    - Included fallback handling for NotificationManager errors.

- **Bug Fixes/Improvements:** 
  - Improved user feedback through system notifications.
  - Enhanced reliability of notification handling with fallback mechanisms for notification managers.

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [X] This PR resolves #238
- [X] Tests have been added for new functionality
- [X] Screenshots or UI changes have been attached
- [X] Documentation has been updated

##  Testing

- **Manual Testing:**
    - [X] Tested on emulator / physical device.
- **Unit Tests**
    - [X] Added units tests for this
- **UI Tests**
    -  [X] Added ui tests for this

##  Related Issues/Tickets

Closes #238 

## Screenshots

Screenshot 2024-11-28 at 19 26 51
Screenshot 2024-11-28 at 19 27 09
Screenshot 2024-11-28 at 19 55 16

### CODE REVIEW

**Review by apana03 — CHANGES_REQUESTED** [review]
Great job implementing this complex issue! Also thank you for the very detailed documentation it was very easy to understand your code and you followed all the relevant best practices. Only had minor comments regarding the code. However I see there is no tests for this PR, they should be added as this is a complex feature and all the cases regarding permissions must be taken into account. If you don't plan to add tests in this PR please open another issue.

File: app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          PendingIntent.getActivity(
+              context, 0, it, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
+        }
+
+    val builder =
+        NotificationCompat.Builder(context, CHANNEL_ID)
+            .setSmallIcon(iconResId)
+            .setContentTitle(title)
+            .setContentText(text)
+            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt]: Thank you for the excellent suggestion! I’ve just implemented the feature to allow the notification priority to be passed as a parameter in showNotification. While the default remains PRIORITY_DEFAULT, this enhancement adds flexibility for use cases requiring higher priority levels, such as PRIORITY_HIGH for critical notifications.

File: app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  fun createNotificationChannel(context: Context) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      val importance = NotificationManager.IMPORTANCE_DEFAULT
+      val channel =
+          NotificationChannel(CHANNEL_ID, CHANNEL_NAME, importance).apply {
+            description = CHANNEL_DESCRIPTION
+          }
+      val notificationManager: NotificationManager =
+          context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+      notificationManager.createNotificationChannel(channel)
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt]: Great observation! I've added a null check for the NotificationManager instance in createNotificationChannel to ensure it handles unexpected scenarios gracefully and prevents potential crashes. Thanks for pointing it out!

File: app/src/main/java/com/android/voyageur/MainActivity.kt [hunk 3]
```diff
@@ -26,30 +32,66 @@ import com.google.android.libraries.places.api.net.PlacesClient
...
+                if (!isConnected) {
+                  val intent =
+                      Intent(this@MainActivity, MainActivity::class.java).apply {
+                        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
+                      }
+                  NotificationHelper.showNotification(
+                      context = this@MainActivity,
+                      notificationId = 1,
+                      title = "No Internet Connection",
+                      text = "Please check your network settings.",
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/MainActivity.kt]: Thanks for highlighting that! I've updated the code to replace all hardcoded strings with references to strings.xml to align with best practices and address the feedback from the last sprint!

File: app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+
+object NotificationHelper {
+  private const val CHANNEL_ID = "voyageur_notifications"
+  private const val CHANNEL_NAME = "Voyageur Notifications"
+  private const val CHANNEL_DESCRIPTION = "Notifications from Voyageur App"
+
+  /** Creates a notification channel for Android O and above. */
+  fun createNotificationChannel(context: Context) {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/ui/notifications/NotificationHelper.kt]: Thank you so much! I’m glad the separation of concerns is clear. Your feedback is greatly appreciated!

**Comment by BogdanCDev** [comment]
The line coverage for the new code in the NotificationHelper is above 80%, as confirmed by Jacoco.

However, the modifications to MainActivity are not included in the coverage reported by SonarCloud. I believe this is due to a specific setting in SonarCloud that excludes coverage for this part of the code. Also, while the changes in MainActivity could theoretically be tested, it would be extremely challenging.

If needed, I can investigate further to clarify the SonarCloud configuration or explore potential test solutions.

**Review by apana03 — APPROVED** [review]
Thank you for taking into account the comments! Also nice job figuring out how to test this new feature as it is not trivial :)

**Pull request merged**